<?xml version="1.0"?>

<robot name="FourRevolute">
  
  <!-- *************************************************** -->
  <!-- ******************** Kinematics ******************* -->
  <!-- *************************************************** -->
  <!-- Note the 'origin' tag in the joints defines the
       shift and reorientation before the continuous joint
       rotates around the 'axis'. -->

  <link name="world"/>

  <joint name="theta1" type="continuous">
    <parent link="world"/>
    <child  link="link1"/>
    <origin xyz="0 0 0.4" rpy="0 0 0"/>
    <axis   xyz="0 0 1"/>
  </joint>

  <link name="link1"/>

  <joint name="theta2" type="continuous">
    <parent link="link1"/>
    <child  link="link2"/>
    <origin xyz="0 0.4 0" rpy="0 0 0"/>
    <axis   xyz="0 1 0"/>
  </joint>

  <link name="link2"/>

  <joint name="theta3" type="continuous">
    <parent link="link2"/>
    <child  link="link3"/>
    <origin xyz="0 0 -0.4" rpy="0 0 0"/>
    <axis   xyz="0 0 -1"/>
  </joint>

  <link name="link3"/>

  <joint name="theta4" type="continuous">
    <parent link="link3"/>
    <child  link="link4"/>
    <origin xyz="0 0.4 0" rpy="0 0 0"/>
    <axis   xyz="0 1 0"/>
  </joint>

  <link name="link4"/>

  <joint name="final" type="fixed">
    <parent link="link4"/>
    <child  link="tip"/>
    <origin xyz="0 0 0" rpy="1.5708 0 3.1416"/>
  </joint>

  <link name="tip"/>


  <!-- *************************************************** -->
  <!-- ******************* RVIZ Colors ******************* -->
  <!-- *************************************************** -->

  <material name="white">    <color rgba="1.00 1.00 1.00 1"/>  </material>
  <material name="gray">     <color rgba="0.60 0.60 0.60 1"/>  </material>
  <material name="table">    <color rgba="0.85 0.77 0.77 1"/>  </material>
  <material name="black">    <color rgba="0.00 0.00 0.00 1"/>  </material>

  <material name="red">      <color rgba="1.00 0.00 0.00 1"/>  </material>
  <material name="green">    <color rgba="0.00 1.00 0.00 1"/>  </material>
  <material name="blue">     <color rgba="0.00 0.00 1.00 1"/>  </material>

  <material name="cyan">     <color rgba="0.00 1.00 1.00 1"/>  </material>
  <material name="magenta">  <color rgba="1.00 0.00 1.00 1"/>  </material>
  <material name="yellow">   <color rgba="1.00 1.00 0.00 1"/>  </material>

  <material name="orange">   <color rgba="1.00 0.65 0.00 1"/>  </material>


  <!-- *************************************************** -->
  <!-- ****** Building Blocks (Visual AND Inertial) ****** -->
  <!-- *************************************************** -->

  <!-- *** Attachment Rod 1, Motor 1, Output Shaft 1 *** -->

  <joint name="attach-rod1" type="fixed">
    <parent link="world"/>
    <child  link="rod1"/>
  </joint>
  <link name="rod1">
    <visual>
      <origin xyz="0 0 0.05" rpy="0 0 0"/>
      <geometry>
	<cylinder length="0.1" radius="0.01"/>
      </geometry>
      <material name="black"/>
    </visual>
    <inertial>
      <origin xyz="0 0 0.05" rpy="0 0 0"/>
      <mass value="0.1"/>
      <inertia
        ixx="0.0001" ixy="0.0" ixz="0.0"
        iyy="0.0001" iyz="0.0"
        izz="0.000005"/>
    </inertial>
  </link>

  <joint name="attach-motor1" type="fixed">
    <parent link="world"/>
    <child  link="motor1"/>
  </joint>
  <link name="motor1">
    <visual>
      <origin xyz="0 0 0.2" rpy="0 0 0"/>
      <geometry>
	<cylinder length="0.2" radius="0.05"/>
      </geometry>
      <material name="gray"/>
    </visual>
    <inertial>
      <origin xyz="0 0 0.2" rpy="0 0 0"/>
      <mass value="1.0"/>
      <inertia
        ixx="0.003" ixy="0.0" ixz="0.0"
        iyy="0.003" iyz="0.0"
        izz="0.001"/>
    </inertial>
  </link>

  <joint name="attach-out1" type="fixed">
    <parent link="link1"/>
    <child  link="out1"/>
  </joint>
  <link name="out1">
    <visual>
      <origin xyz="0 0 -0.05" rpy="0 0 0"/>
      <geometry>
	<cylinder length="0.1" radius="0.01"/>
      </geometry>
      <material name="black"/>
    </visual>
    <inertial>
      <origin xyz="0 0 -0.05" rpy="0 0 0"/>
      <mass value="0.1"/>
      <inertia
        ixx="0.0001" ixy="0.0" ixz="0.0"
        iyy="0.0001" iyz="0.0"
        izz="0.000005"/>
    </inertial>
  </link>

  <!-- *** Attachment Rod 2, Motor 2, Output Shaft 2 *** -->

  <joint name="attach-rod2" type="fixed">
    <parent link="link1"/>
    <child  link="rod2"/>
  </joint>
  <link name="rod2">
    <visual>
      <origin xyz="0 0.05 0" rpy="-1.5708 0 0"/>
      <geometry>
	<cylinder length="0.1" radius="0.01"/>
      </geometry>
      <material name="black"/>
    </visual>
    <inertial>
      <origin xyz="0 0.05 0" rpy="-1.5708 0 0"/>
      <mass value="0.1"/>
      <inertia
        ixx="0.0001" ixy="0.0" ixz="0.0"
        iyy="0.0001" iyz="0.0"
        izz="0.000005"/>
    </inertial>
  </link>

  <joint name="attach-motor2" type="fixed">
    <parent link="link1"/>
    <child  link="motor2"/>
  </joint>
  <link name="motor2">
    <visual>
      <origin xyz="0 0.2 0" rpy="-1.5708 0 0"/>
      <geometry>
	<cylinder length="0.2" radius="0.05"/>
      </geometry>
      <material name="gray"/>
    </visual>
    <inertial>
      <origin xyz="0 0.2 0" rpy="-1.5708 0 0"/>
      <mass value="1.0"/>
      <inertia
        ixx="0.003" ixy="0.0" ixz="0.0"
        iyy="0.003" iyz="0.0"
        izz="0.001"/>
    </inertial>
  </link>

  <joint name="attach-out2" type="fixed">
    <parent link="link2"/>
    <child  link="out2"/>
  </joint>
  <link name="out2">
    <visual>
      <origin xyz="0 -0.05 0" rpy="-1.5708 0 0"/>
      <geometry>
	<cylinder length="0.1" radius="0.01"/>
      </geometry>
      <material name="black"/>
    </visual>
    <inertial>
      <origin xyz="0 -0.05 0" rpy="-1.5708 0 0"/>
      <mass value="0.1"/>
      <inertia
        ixx="0.0001" ixy="0.0" ixz="0.0"
        iyy="0.0001" iyz="0.0"
        izz="0.000005"/>
    </inertial>
  </link>

  <!-- *** Attachment Rod 3, Motor 3, Output Shaft 3 *** -->

  <joint name="attach-rod3" type="fixed">
    <parent link="link2"/>
    <child  link="rod3"/>
  </joint>
  <link name="rod3">
    <visual>
      <origin xyz="0 0 -0.05" rpy="0 0 0"/>
      <geometry>
	<cylinder length="0.1" radius="0.01"/>
      </geometry>
      <material name="black"/>
    </visual>
    <inertial>
      <origin xyz="0 0 -0.05" rpy="0 0 0"/>
      <mass value="0.1"/>
      <inertia
        ixx="0.0001" ixy="0.0" ixz="0.0"
        iyy="0.0001" iyz="0.0"
        izz="0.000005"/>
    </inertial>
  </link>

  <joint name="attach-motor3" type="fixed">
    <parent link="link2"/>
    <child  link="motor3"/>
  </joint>
  <link name="motor3">
    <visual>
      <origin xyz="0 0 -0.2" rpy="0 0 0"/>
      <geometry>
	<cylinder length="0.2" radius="0.05"/>
      </geometry>
      <material name="gray"/>
    </visual>
    <inertial>
      <origin xyz="0 0 -0.2" rpy="0 0 0"/>
      <mass value="1.0"/>
      <inertia
        ixx="0.003" ixy="0.0" ixz="0.0"
        iyy="0.003" iyz="0.0"
        izz="0.001"/>
    </inertial>
  </link>

  <joint name="attach-out3" type="fixed">
    <parent link="link3"/>
    <child  link="out3"/>
  </joint>
  <link name="out3">
    <visual>
      <origin xyz="0 0 0.05" rpy="0 0 0"/>
      <geometry>
	<cylinder length="0.1" radius="0.01"/>
      </geometry>
      <material name="black"/>
    </visual>
    <inertial>
      <origin xyz="0 0 0.05" rpy="0 0 0"/>
      <mass value="0.1"/>
      <inertia
        ixx="0.0001" ixy="0.0" ixz="0.0"
        iyy="0.0001" iyz="0.0"
        izz="0.000005"/>
    </inertial>
  </link>

  <!-- *** Attachment Rod 4, Motor 4, Output Shaft 4 *** -->

  <joint name="attach-rod4" type="fixed">
    <parent link="link3"/>
    <child  link="rod4"/>
  </joint>
  <link name="rod4">
    <visual>
      <origin xyz="0 0.05 0" rpy="-1.5708 0 0"/>
      <geometry>
	<cylinder length="0.1" radius="0.01"/>
      </geometry>
      <material name="black"/>
    </visual>
    <inertial>
      <origin xyz="0 0.05 0" rpy="-1.5708 0 0"/>
      <mass value="0.1"/>
      <inertia
        ixx="0.0001" ixy="0.0" ixz="0.0"
        iyy="0.0001" iyz="0.0"
        izz="0.000005"/>
    </inertial>
  </link>

  <joint name="attach-motor4" type="fixed">
    <parent link="link3"/>
    <child  link="motor4"/>
  </joint>
  <link name="motor4">
    <visual>
      <origin xyz="0 0.2 0" rpy="-1.5708 0 0"/>
      <geometry>
	<cylinder length="0.2" radius="0.05"/>
      </geometry>
      <material name="gray"/>
    </visual>
    <inertial>
      <origin xyz="0 0.2 0" rpy="-1.5708 0 0"/>
      <mass value="1.0"/>
      <inertia
        ixx="0.003" ixy="0.0" ixz="0.0"
        iyy="0.003" iyz="0.0"
        izz="0.001"/>
    </inertial>
  </link>

  <joint name="attach-out4" type="fixed">
    <parent link="link4"/>
    <child  link="out4"/>
  </joint>
  <link name="out4">
    <visual>
      <origin xyz="0 -0.05 0" rpy="-1.5708 0 0"/>
      <geometry>
	<cylinder length="0.1" radius="0.01"/>
      </geometry>
      <material name="black"/>
    </visual>
    <inertial>
      <origin xyz="0 -0.05 0" rpy="-1.5708 0 0"/>
      <mass value="0.1"/>
      <inertia
        ixx="0.0001" ixy="0.0" ixz="0.0"
        iyy="0.0001" iyz="0.0"
        izz="0.000005"/>
    </inertial>
  </link>

  <!-- *** Hand: Left Finger, Palm Bar, Right Finger *** -->

  <joint name="attach-leftfinger" type="fixed">
    <parent link="tip"/>
    <child  link="leftfinger"/>
  </joint>
  <link name="leftfinger">
    <visual>
      <origin xyz="0.05 0 0.025" rpy="0 0 0"/>
      <geometry>
	<cylinder length="0.05" radius="0.01"/>
      </geometry>
      <material name="black"/>
    </visual>
    <inertial>
      <origin xyz="0.05 0 0.025" rpy="0 0 0"/>
      <mass value="0.05"/>
      <inertia
        ixx="0.0000125" ixy="0.0" ixz="0.0"
        iyy="0.0000125" iyz="0.0"
        izz="0.0000025"/>
    </inertial>
  </link>

  <joint name="attach-rightfinger" type="fixed">
    <parent link="tip"/>
    <child  link="rightfinger"/>
  </joint>
  <link name="rightfinger">
    <visual>
      <origin xyz="-0.05 0 0.025" rpy="0 0 0"/>
      <geometry>
	<cylinder length="0.05" radius="0.01"/>
      </geometry>
      <material name="black"/>
    </visual>
    <inertial>
      <origin xyz="-0.05 0 0.025" rpy="0 0 0"/>
      <mass value="0.05"/>
      <inertia
        ixx="0.0000125" ixy="0.0" ixz="0.0"
        iyy="0.0000125" iyz="0.0"
        izz="0.0000025"/>
    </inertial>
  </link>

  <joint name="attach-palm" type="fixed">
    <parent link="tip"/>
    <child  link="palm"/>
  </joint>
  <link name="palm">
    <visual>
      <origin xyz="0 0 0" rpy="0 1.5708 0"/>
      <geometry>
	<cylinder length="0.10" radius="0.01"/>
      </geometry>
      <material name="black"/>
    </visual>
    <inertial>
      <origin xyz="0 0 0" rpy="0 1.5708 0"/>
      <mass value="0.1"/>
      <inertia
        ixx="0.0001" ixy="0.0" ixz="0.0"
        iyy="0.0001" iyz="0.0"
        izz="0.000005"/>
    </inertial>
  </link>
  

  <!-- *************************************************** -->
  <!-- ******************* Gazebo Colors ***************** -->
  <!-- *************************************************** -->
  
  <gazebo reference="rod1">   <material>Gazebo/Black</material> </gazebo>
  <gazebo reference="motor1"> <material>Gazebo/Gray</material>  </gazebo>
  <gazebo reference="out1">   <material>Gazebo/Black</material> </gazebo>

  <gazebo reference="rod2">   <material>Gazebo/Black</material> </gazebo>
  <gazebo reference="motor2"> <material>Gazebo/Gray</material>  </gazebo>
  <gazebo reference="out2">   <material>Gazebo/Black</material> </gazebo>

  <gazebo reference="rod3">   <material>Gazebo/Black</material> </gazebo>
  <gazebo reference="motor3"> <material>Gazebo/Gray</material>  </gazebo>
  <gazebo reference="out3">   <material>Gazebo/Black</material> </gazebo>

  <gazebo reference="rod4">   <material>Gazebo/Black</material> </gazebo>
  <gazebo reference="motor4"> <material>Gazebo/Gray</material>  </gazebo>
  <gazebo reference="out4">   <material>Gazebo/Black</material> </gazebo>

  <gazebo reference="rod4">   <material>Gazebo/Black</material> </gazebo>
  <gazebo reference="motor4"> <material>Gazebo/Gray</material>  </gazebo>
  <gazebo reference="out4">   <material>Gazebo/Black</material> </gazebo>

  <gazebo reference="leftfinger">  <material>Gazebo/Black</material> </gazebo>
  <gazebo reference="rightfinger"> <material>Gazebo/Black</material> </gazebo>
  <gazebo reference="palm">        <material>Gazebo/Black</material> </gazebo>


</robot>
